### PR TITLE
[Fix #51164] `Model.query_constraints` with single non-primary-key column raises incorrect error

### LINKED ISSUE #51164: `Model.query_constraints` with a single non-primary-key column raises a confusing error
### Steps to reproduce

Using `Model.query_constraints` with a single column that's not the primary key raises an error meant for passing more than 2 columns. This issue has existed since these raises were originally added in #49088.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", path: "../rails"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :organisations, force: true do |t|
  end

  create_table :employees, force: true do |t|
    t.integer :organisation_id
  end
end

class Organisation < ActiveRecord::Base
  has_many :employees
end

class Employee < ActiveRecord::Base
  query_constraints :organisation_id

  belongs_to :organisation
end

class BugTest < Minitest::Test
  def test_query_constraints_with_single_non_primary_key_column
    org = Organisation.create!
    org.employees << Employee.create!

    assert_equal 1, org.employees.count  # ❌ raises as it should, but with a confusing message
  end
end

```

### Expected behavior

It turns out there is [already an appropriate error for this case](https://github.com/rails/rails/blob/9f1dec2ea5155205a880f6e6e232cf9ea6da2d8c/activerecord/lib/active_record/reflection.rb#L799-L806). We never reach it because of the incorrect condition for the error before. I would expect this to get raised:

```ruby
Error:
BugTest#test_association_stuff:
ArgumentError: The query constraints on the `Employee` model does not include the primary key so Active Record is unable to derive the foreign key constraints for the association. You need to explicitly define the query constraints for this association.
    test_2.rb:45:in `test_query_constraints_with_single_non_primary_key_column'
```

### Actual behavior

Raises this error:

```ruby
Error:
BugTest#test_association_stuff:
ArgumentError: The query constraints list on the `Employee` model has more than 2 attributes. Active Record is unable to derive the query constraints for the association. You need to explicitly define the query constraints for this association.
    test_2.rb:45:in `test_query_constraints_with_single_non_primary_key_column'
```

### System configuration
**Rails version**: 7.1.3, main

**Ruby version**: 3.3.0

### PR DESCRIPTION
**Pull request by joshuay03**
### Motivation / Background

Fixes #51164.

### Detail

This Pull Request changes `ActiveRecord::Reflection::AssociationReflection#derive_fk_query_constraints` to raise an existing more appropriate error if `Model.query_constraints` is given a single non-primary-key column argument.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/test/cases/associations_test.rb [hunk 1]
```diff
@@ -368,7 +368,22 @@ def test_assign_composite_foreign_key_belongs_to_association
...
+    original = Sharded::BlogPost.instance_variable_get(:@query_constraints_list)
+    Sharded::BlogPost.query_constraints :title
+    Sharded::BlogPost.has_many :comments_without_query_constraints, primary_key: [:blog_id, :id], class_name: "Comment"
+    blog_post = sharded_blog_posts(:great_post_blog_one)
+
+    error = assert_raises ArgumentError do
+      blog_post.comments_without_query_constraints.to_a
+    end
+
+    assert_equal "The query constraints on the `Sharded::BlogPost` model does not include the primary key so Active Record is unable to derive the foreign key constraints for the association. You need to explicitly define the query constraints for this association.", error.message
```
Comment (joshuay03) [inline activerecord/test/cases/associations_test.rb:381]: Before patch:

Screenshot 2024-02-22 at 11 59 06 PM

**Review by eileencodes — CHANGES_REQUESTED** [review]
Can you fix the tests? You need to use a different name for the `has_many`, the redefined ones need to be unique per test.

**Comment by joshuay03** [comment]
> Can you fix the tests? You need to use a different name for the `has_many`, the redefined ones need to be unique per test.

@eileencodes Done, thanks for taking a look! The current CI failure seems unrelated.

**Review by eileencodes — APPROVED** [review]

**Comment by eileencodes** [comment]
Thanks!

Backported to 7-1-stable in 6f0c3a4721d2ff0190a17d687198c9b84decf2cf

**Pull request merged**